### PR TITLE
[GH-436] Add integration tests for missing User API endpoints

### PR DESCRIPTION
**Pull request by yedamao**
#### Summary

Add integration tests for missing User API endpoints

#### Ticket Link

Fixes https://github.com/mattermost/focalboard/issues/436

### CODE REVIEW

**Comment by harshilsharma63** [comment]
Hi @yedamao ! Thanks for the PR.

The PR is in the draft right now, so I assume you're working on adding some more tests?

**Comment by harshilsharma63** [comment]
Can you also explain a bit about your implementation? For example, I see you added a `native` authentication mode. I'm not sure what's that for.

**Comment by yedamao** [comment]
The original test helper server run in single user mode can’t perform register/login or change password operations, so I add a `SetupTestHelperWithouToken`(maybe need a better name) method to setup a test server not in single user mode.

**Comment by yedamao** [comment]
> Can you also explain a bit about your implementation? For example, I see you added a `native` authentication mode. I'm not sure what's that for.

emm... I think `native` is designed auth type.


login:
```go
  authService := user.AuthService
  if authService == "" {
    authService = "native"
  }

  session := model.Session{
    ID:          uuid.New().String(),
    Token:       uuid.New().String(),
    UserID:      user.ID,
    AuthService: authService,
    Props:       map[string]interface{}{},
  }
  err := a.store.CreateSession(&session)
  if err != nil {
    return "", errors.Wrap(err, "unable to create session")
  }
```

in attachSesstion:
```go
    session, err := a.app.GetSession(token)
    if err != nil {
      if required {
        a.errorResponse(w, r.URL.Path, http.StatusUnauthorized, "", err)
        return
      }

      handler(w, r)
      return
    }

    authService := session.AuthService
    if authService != a.authService {
      a.logger.Error(`Session authService mismatch`,
        mlog.String("sessionID", session.ID),
        mlog.String("want", a.authService),
        mlog.String("got", authService),
      )
      a.errorResponse(w, r.URL.Path, http.StatusUnauthorized, "", err)
      return
    }
```

if `config.AuthMode` is empty,when user login the session's authService field will be set as `"native"`, and  the check in attachSesstion will not pass.

**Comment by yedamao** [comment]
Maybe I understand it wrong, I'd like to have your suggestions.

**Comment by hahmadia** [comment]
Hi @yedamao , 

I marked this as ready for review. Is that correct? Are you ready to have staff review your PR or are you still working on it?
Thanks

**Comment by yedamao** [comment]
@harshilsharma63 ok it's ready

File: server/integrationtests/user_test.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+			Password: password,
+		}
+		data, resp := th.Client.Login(loginRequest)
+		require.NoError(t, resp.Error)
+		require.NotNil(t, data)
+		require.NotNil(t, data.Token)
+	})
+}
+
+func TestGetUserMe(t *testing.T) {
```
Comment (hahmadia) [inline server/integrationtests/user_test.go:78]:
```suggestion
func TestGetMe(t *testing.T) {
```

File: server/client/client.go [hunk 2]
```diff
@@ -210,3 +218,90 @@ func (c *Client) PostSharing(sharing model.Sharing) (bool, *Response) {
...
+	}
+
+	if data.Token != "" {
+		c.Token = data.Token
+	}
+
+	return data, BuildResponse(r)
+}
+
+func (c *Client) GetUserMeRoute() string {
```
Comment (hahmadia) [inline server/client/client.go:259]:
```suggestion
func (c *Client) GetMeRoute() string {
```

File: server/client/client.go [hunk 3]
```diff
@@ -210,3 +218,90 @@ func (c *Client) PostSharing(sharing model.Sharing) (bool, *Response) {
...
+	return "/users/me"
+}
+
+func (c *Client) GetUserMe() (*model.User, *Response) {
```
Comment (hahmadia) [inline server/client/client.go:263]:
```suggestion
func (c *Client) GetMe() (*model.User, *Response) {
```

File: server/integrationtests/user_test.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+		}
+		data, resp := th.Client.Login(loginRequest)
+		require.NoError(t, resp.Error)
+		require.NotNil(t, data)
+		require.NotNil(t, data.Token)
+
+		// get user me
+		me, resp := th.Client.GetUserMe()
+		require.NoError(t, resp.Error)
+		require.NotNil(t, me)
```
Comment (hahmadia) [inline server/integrationtests/user_test.go:114]: If it's okay, could you just add a few checks to make sure the data in `me` matches the above data that you used to register the user (you are really getting back your own data and not someone elses) like you've done in other tests.

**Review by hahmadia — CHANGES_REQUESTED** [review]
Looks great for the most part. Just some minor changes I want to suggest. 
Thanks

**Comment by hahmadia** [comment]
If it isn't too much extra work, could we also get a test for `apiv1.HandleFunc("/workspaces/{workspaceID}/{rootID}/files", a.sessionRequired(a.handleUploadFile)).Methods("POST")` as it's part of the USER API's endpoint that seems to have been missed here. Thank you a lot :)

File: server/integrationtests/user_test.go (hunk 4 above)
Comment (yedamao) [reply server/integrationtests/user_test.go:114]: Add some checks, done

**Comment by yedamao** [comment]
@hahmadia I have fixed the minor changes in comments.

**Comment by yedamao** [comment]
@hahmadia I am working on upload file api

**Comment by hahmadia** [comment]
> @hahmadia I am working on upload file api

Alright Thanks @yedamao . Let me know when it's complete and I can re-review it. Appreciate your hard work!

**Comment by yedamao** [comment]
@hastadhana I have done the workspace upload file api.

**Review by harshilsharma63 — APPROVED** [review]
LGTM

**Review by hahmadia — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by yedamao** [comment]
ping...

**Comment by hahmadia** [comment]
could you please resolve the conflict your branch has with master? Thanks

**Comment by hahmadia** [comment]
@yedamao

**Comment by yedamao** [comment]
@hahmadia ok, the conflict has been resolved.

**Pull request merged**